### ISSUE TITLE
Delete Stream and recreate with same streamname causes exception

### ISSUE THREAD

**Issue by addprs** [pravega/pravega#3057]
**Problem description**
Created stream and deleted the stream, but when we try to create the stream name with the deleted stream name exception occurs:

java.lang.NullPointerException
	at io.pravega.client.stream.impl.StreamSegments.getSegmentForKey(StreamSegments.java:67)
	at io.pravega.client.stream.impl.SegmentSelector.getSegmentForEvent(SegmentSelector.java:78)
	at io.pravega.client.stream.impl.SegmentSelector.getSegmentOutputStreamForKey(SegmentSelector.java:69)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.resend(EventStreamWriterImpl.java:180)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$null$1(EventStreamWriterImpl.java:155)
	at io.pravega.common.util.Retry$RetryAndThrowBase.run(Retry.java:173)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$handleLogSealed$2(EventStreamWriterImpl.java:142)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

**Problem location**
How to reproduce it :
    public static void main(String[] args) throws Exception {
        String testScope = "testscope";
        String testStream1 = "teststream1";
       ClientConfig clientConfig = ClientConfig.builder()
                .controllerURI(new URI("tcp://localhost:9090"))
                .build();
       StreamManager streamManager;
        StreamConfiguration streamConfig = StreamConfiguration.builder()
                .scalingPolicy(ScalingPolicy.fixed(1))
                .build();
        streamManager = StreamManager.create(clientConfig);
        streamManager.createScope(testScope);
        streamManager.createStream(testScope, testStream1, streamConfig);
        final EventStreamWriter<String> writer= ClientFactories.get(SCOPE).createEventWriter(
                STREAM1,
                new JavaSerializer<>(),
                EventWriterConfig.builder().build());

        final String data1 = "testdata1";
        final String data2 = "testdata2";
        writer.writeEvent(data1);
        writer.writeEvent(data2);
        TimeUnit.MINUTES.sleep(1);
        streamManager.sealStream(testScope, testStream1);
        streamManager.deleteStream(testScope, testStream1);
        TimeUnit.MINUTES.sleep(1);
        streamManager.createStream(testScope, testStream1, streamConfig);
        writer.writeEvent(data1);
        writer.writeEvent(data2);
    }

**Comment by shiveshr** [comment]
@addprs On which pravega build have you noticed this error?
I tried reproducing against latest master but couldnt reproduce it.

**Comment by RaulGracia** [comment]
This was indeed a bug, but it was fixed in this PR #2800. I noticed that this PR was not cherry-picked to `r0.3`, so this means that this problem is still present in all the `0.3.X` versions of Pravega. It is fixed in the new branch `r0.4`.

**Comment by addprs** [comment]
@shiveshr  we are using 0.3.2 release

**Comment by addprs** [comment]
@RaulGracia with the next release this issue should be gone then?

**Comment by RaulGracia** [comment]
@addprs yes, it should.

**Comment by addprs** [comment]
@RaulGracia updated pravega to r0.4-0.4.0-2005.9df106d  that would be the candidate release.Still i see the same exception, Can you please look over once.Thank you.Or should i be using any later release?

**Comment by RaulGracia** [comment]
@addprs I reproduced your issue and understand the root cause of the problem. First, note that the ability to re-create a Stream has been addressed in PR #2800. We have [integration tests](https://github.com/pravega/pravega/blob/master/test/integration/src/test/java/io/pravega/test/integration/StreamRecreationTest.java) and [system tests](https://github.com/pravega/pravega/blob/master/test/system/src/test/java/io/pravega/test/system/StreamsAndScopesManagementTest.java) that write/read data to/from a Stream, re-create the Stream, and write/read data again to/from the Stream.

However, what your piece of code is testing is not only that a Stream can be recreated, but that the "same writer" that was opened for the previous Stream can work on the new one. At first glace, the failure sounds reasonable: you are creating a _new stream with the same name than a previously deleted one_, but this does not mean that they are the same Stream. Therefore, if you want to write data to the re-created Stream you will need to instantiate a new writer. 

If you look at the tests that cover this issue, in none of them we attempt to reuse the same writer instance across Streams.
